HDDS-12541. Change ContainerID to value-based.

szetszwo opened the pull request:
## What changes were proposed in this pull request?

In order to change ContainerID to be a [value-based class](https://docs.oracle.com/javase/8/docs/api/java/lang/doc-files/ValueBased.html), we must enforce factory method (i.e. disallow directly calling constructors).

We also improve its efficiency here:
- Cache proto and hash.
- Avoid apache.commons.lang3.

## What is the link to the Apache JIRA

HDDS-12541

## How was this patch tested?

Updating existing tests.

adoroszlai commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/container/ReplicationManagerReport.java:
```diff
@@ -154,10 +154,6 @@ public Map<String, Long> getStats() {
     return result;
   }
 
-  /**
-   * Return a map of all samples, with the stat as the key and the samples
-   * for the stat as a List of Long.
-   */
```
This change doesn't seem to be related.

adoroszlai commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerID.java:
```diff
@@ -46,16 +45,20 @@ public static Codec<ContainerID> getCodec() {
   }
 
   private final long id;
+  private final HddsProtos.ContainerID proto;
+  private final int hash;
```
We can add `@Immutable` on the class.  (Only if you update the patch anyway.)

(`import net.jcip.annotations.Immutable;`)

adoroszlai reviewed (commented):
Thanks @szetszwo for the patch.

nandakumar131 commented:
Thanks @szetszwo for the improvement.
Since `ContainerID` is immutable and all the Container IDs are kept in SCM's memory, how about adding a cache layer inside `ContainerID` class and return from cache if the `ContainerID` is already present in cache rather than creating new one?

This will avoid re-calculation of `hash` and `proto` conversion and also help (though very little) with SCM's memory consumption.

adoroszlai commented:
I guess instance caching would be a follow-up.

szetszwo commented:
@adoroszlai , thanks for the comments.  I have addressed them.

@nandakumar131 , caching is a good idea. Let's do it separately since this change is already big.

For the proto and the hash, we could compute them lazy.  Updated the patch.

adoroszlai commented:
Thanks @szetszwo for the patch, @nandakumar131 for the review.
